CORE-13827 - Sync start offsets from source to shadow

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR implements automatic prefix-truncation based on updates to the source parititon's log start offset.  As that changes, the Shadow Link will automatically prefix-truncate the shadow partition.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
cl: Add prefix_truncate and start_offset() to data_sink

These methods will be used to perform prefix truncation based on the
start offset of the source partition.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/rep: Added method to truncate shadow partition

When partition replicator receives a new batch, it will check the
offsets of the source and shadow partitions to see if the shadow
partition needs to be truncated.  If the source start offset is greater
than the shadow partitions source start AND it is less than the shadow
paritions HWM, then it will truncate the shadow partition.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Add prefix trimming tests

Signed-off-by: Michael Boquard <michael@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--35a48565-dc66-4137-930b-bc0d379af216-->
<!--TEST RESULTS-->
<!--194681a5-1202-4500-89a1-6cfc29b9bf0a-->
<!--TEST RESULTS-->
<!--2604829a-aa40-4034-b272-993d9fde47c9-->
<!--TEST RESULTS-->
<!--d589ef39-20cd-4960-be83-c4f47fa59033-->
<!--TEST RESULTS-->
<!--ca230a0f-0967-46f1-ac38-7e8daa098234-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74647">build#74647</a></summary>


| test_class                 | test_method                                             | test_arguments                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                   |
|:---------------------------|:--------------------------------------------------------|:----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest      | test_creating_and_listing_migrations                    | null                                          | integration | https://buildkite.com/redpanda/redpanda/builds/74647#019a0935-8e29-485e-8b4c-fa403e1a436b | FLAKY         | 18/21    | upstream reliability is '99.81343283582089'. current run reliability is '85.71428571428571'. drift is 14.09915 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations                       |
| NodesDecommissioningTest   | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": false, "delete_topic": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74647#019a0935-8e26-4d55-bb01-f53b66047d98 | FLAKY         | 19/21    | upstream reliability is '99.0909090909091'. current run reliability is '90.47619047619048'. drift is 8.61472 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| ShadowLinkingRandomOpsTest | test_node_operations                                    | {"failures": false}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74647#019a0935-8e2a-4fc0-9a65-e88dfd1116c0 | FLAKY         | 5/21     | upstream reliability is '39.75155279503105'. current run reliability is '23.809523809523807'. drift is 15.94203 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest | test_node_operations                                    | {"failures": false}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74647#019a0936-60b3-48dc-a77b-faea4cd9d328 | FLAKY         | 9/21     | upstream reliability is '39.75155279503105'. current run reliability is '42.857142857142854'. drift is -3.10559 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
| ShadowLinkingRandomOpsTest | test_node_operations                                    | {"failures": true}                            | integration | https://buildkite.com/redpanda/redpanda/builds/74647#019a0935-8e2b-4ac7-95a6-fe53f5cb3414 | FLAKY         | 1/21     | upstream reliability is '32.58426966292135'. current run reliability is '4.761904761904762'. drift is 27.82236 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations                                  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74795">build#74795</a></summary>


| test_class                 | test_method          | test_arguments      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                    | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:--------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74795#019a11a6-0ad9-439a-b376-ee34bd54db20 | FLAKY         | 5/21     | upstream reliability is '36.795252225519285'. current run reliability is '23.809523809523807'. drift is 12.98573 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74795#019a11a7-207f-42a2-98c9-77fc470f1bcb | FLAKY         | 3/21     | upstream reliability is '36.795252225519285'. current run reliability is '14.285714285714285'. drift is 22.50954 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74795#019a11a6-0ada-4adb-a123-2499f126c753 | FLAKY         | 3/21     | upstream reliability is '31.442463533225286'. current run reliability is '14.285714285714285'. drift is 17.15675 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74795#019a11a7-2080-450a-a3e6-4bb4687c034d | FLAKY         | 6/21     | upstream reliability is '52.185089974293064'. current run reliability is '28.57142857142857'. drift is 23.61366 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74850">build#74850</a></summary>


| test_class                   | test_method                          | test_arguments      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:-----------------------------|:-------------------------------------|:--------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests         | test_rapid_shadow_link_toggling      | null                | integration | https://buildkite.com/redpanda/redpanda/builds/74850#019a132f-bd85-4465-8861-bf8e200b5d7b | FLAKY         | 16/21    | upstream reliability is '93.26633165829146'. current run reliability is '76.19047619047619'. drift is 17.07586 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling       |
| ShadowLinkUpdateBrokersTests | test_update_brokers                  | null                | integration | https://buildkite.com/redpanda/redpanda/builds/74850#019a132f-bd8b-4ab2-aac5-e7302565b3c7 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkUpdateBrokersTests&test_method=test_update_brokers           |
| DataMigrationsApiTest        | test_creating_and_listing_migrations | null                | integration | https://buildkite.com/redpanda/redpanda/builds/74850#019a1336-ac3c-40a9-a015-0b99438a6fdf | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| ShadowLinkingRandomOpsTest   | test_node_operations                 | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74850#019a132f-bd8d-47bd-885f-9abe07bba14e | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations            |
| ShadowLinkingRandomOpsTest   | test_node_operations                 | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74850#019a1336-ac3e-4d53-bfa3-f5e5fc049ad3 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations            |
| ShadowLinkingRandomOpsTest   | test_node_operations                 | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74850#019a132f-bd83-42f9-9da0-e174c660ab01 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations            |
| ShadowLinkingRandomOpsTest   | test_node_operations                 | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74850#019a1336-ac36-4e5f-99da-d5d13f6e7b2f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations            |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74879">build#74879</a></summary>


| test_class                 | test_method          | test_arguments      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:--------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74879#019a13d2-f36a-4a3b-afd0-64cb7b648772 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": false} | integration | https://buildkite.com/redpanda/redpanda/builds/74879#019a13d9-c03b-4d2a-8d63-d58895da4bec | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74879#019a13d2-f36a-4a3b-afd0-64cb7b648772 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/74879#019a13d9-c03b-4d2a-8d63-d58895da4bec | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74890">build#74890</a></summary>


| test_class                      | test_method              | test_arguments                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      | test_history                                                                                                                                                           |
|:--------------------------------|:-------------------------|:--------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74890#019a1481-02b3-4aa1-9347-bef61db679f8 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<!--/body-->

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -208,6 +208,7 @@ ss::future<> partition_replicator::fetch_and_replicate() {
         while (!_gate.is_closed() && !as.abort_requested()) {
             auto inflight_units = co_await ss::get_units(_max_requests, 1, as);
             auto data = co_await _source->fetch_next(as);
+            co_await maybe_synchronize_start_offsets();
```
if there is no new data, this would never get invoked?

bharathv replied:
Also, this in the hot path of replication, perhaps there should be a separate `void sink::notify_start_offset_change(new_offset);` that reconciles the start_offset in the background?

michael-redpanda replied:
> if there is no new data, this would never get invoked?

I can't remember if @joe-redpanda 's changes will allow us to enqueue empty batches with just offset updates or not since it's possible to get a fetch response with no data but updated offsets.

michael-redpanda replied:
> Also, this in the hot path of replication, perhaps there should be a separate void sink::notify_start_offset_change(new_offset); that reconciles the start_offset in the background?

Yeah the trick will be to not build up a huge list of truncations that will effectively be no-ops.

michael-redpanda replied:
Ok I think I addressed this, lmk what you think of my approach

joe-redpanda replied:
I'll have it allow empty batches

bharathv commented on src/v/cluster_link/replication/deps.h:
```diff
@@ -49,6 +49,13 @@ class data_sink {
+    virtual ss::future<std::error_code> prefix_truncate(
+      kafka::offset truncation_offset, ss::lowres_clock::time_point deadline)
+      = 0;
+
```
I'm wondering if this should just be a property of the implementation vs putting it in the API.

IoannisRP commented on src/v/cluster_link/service.cc:
```diff
@@ -492,6 +492,53 @@ class local_partition_sink : public replication::data_sink {
+        if (model_truncation_offset > local_kafka_start_offset) {
+            rp_truncate_offset = translator->to_log_offset(
+              model_truncation_offset);
+        }
```
nit: due to the translator interface, here you end up with a `model::offset` variable called `..._kafka_...`. While it is correct, i find it confusing. I think something like the following is easier to parse for the uninitiated (i.e. me...)
```suggestion
        ::model::offset rp_truncate_offset{};
        auto local_kafka_start_offset = model::offset_cast(translator->from_log_offset(
          _partition->raft_start_offset()));
        if (truncation_offset > local_kafka_start_offset) {
            rp_truncate_offset = translator->to_log_offset(
              kafka::offset_cast(truncation_offset));
        }
```

michael-redpanda replied:
FYI I am attempting to follow what https://github.com/redpanda-data/redpanda/blob/2a4f45610fb89d8f0a5d93eb972dc60be4bb9647/src/v/kafka/data/replicated_partition.cc#L536-L586 does

IoannisRP replied:
the difference that i see is that truncation offset comes as `model::offset` there. So it keeps everything in model::offset till it calls the the partition `prefix_truncate`.

Here you have the input in `kafka::offset`.

It's just a comment for readability. I think the code is correct, so either is find.

michael-redpanda replied:
Yes, the offset for the `delete_records` API is read off the wire as a `model::offset`.  All of the replication `deps` use `kafka::offset`.  So I think it's equivalent in this situation

IoannisRP commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -244,4 +245,62 @@ ss::future<> partition_replicator::fetch_and_replicate() {
+
+    auto source_start_offset = source_offsets->source_start_offset;
+
+    if (source_start_offset <= shadow_partition_start_offset) {
+        vlog(
+          _log.trace,
+          "Source and shadow partition start offsets are in sync: source: {}, "
+          "shadow: {}",
+          source_start_offset,
+          shadow_partition_start_offset);
+        co_return;
+    }
```
question:

Under what scenario do we end up with `source_start_offset < shadow_partition_start_offset`?

michael-redpanda replied:
If we permit users to do prefix truncation on the shadow cluster directly.  Right now we don't but not sure if in the future we wil.

michael-redpanda commented:
Force push:

* Moved prefix truncate out of hot path and used work_queue

michael-redpanda commented:
Force push:

* Fixed test bug

joe-redpanda commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -1277,6 +1277,110 @@ def test_replication_with_truncated_topic(self):
+                num_parts += 1
+                if (part.high_watermark or 0) < offset:
+                    return False
+            return num_parts == 5 and True
```
what is the purpose of 'and true'?
is this not the same as just returning num_parts == 5

joe-redpanda replied:
~~also where is 5 coming from? topic spec seems to be setting the topic up with one partition?~~

ah, this is only getting used in the test below.
can replication partition count be a parameter instead of hardcoded?

michael-redpanda replied:
> what is the purpose of 'and true'?

I... am going to blame AI auto complete on that (otherwise man that is some hilarious boolean algebra going on there)

michael-redpanda replied:
> can replication partition count be a parameter instead of hardcoded?

absolutely

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -208,6 +214,7 @@ ss::future<> partition_replicator::fetch_and_replicate() {
         while (!_gate.is_closed() && !as.abort_requested()) {
             auto inflight_units = co_await ss::get_units(_max_requests, 1, as);
             auto data = co_await _source->fetch_next(as);
+            maybe_synchronize_start_offset();
```
lost the original comment chain but AFAICT fetch_next() doesn't return if there no new data. We need something external to invoke the function on start offset outside of this replicator loop.  I  _think_ it is reasonable to have a timer in link_replication_mgr that periodically sweeps through all replicators and synchronizes start offset if something changed.

michael-redpanda replied:
True - though I do think triggering it as part of the synchronize is still a good thing to do.  So have both?

bharathv replied:
makes sense.

michael-redpanda replied:
>  I think it is reasonable to have a timer in link_replication_mgr that periodically sweeps through all replicators and synchronizes start offset if something changed.

Any opposition to just doing it directly in partition_replicator?

bharathv replied:
I think thats too many timers and overhead IMO.

michael-redpanda replied:
Ok gotcha, yeah that's a good point

michael-redpanda replied:
Ok lmk what you think of the new new idea

michael-redpanda commented:
Force push:

* Addressed PR comments
* Partition replicator will now periodically see if start offset needs sinking

joe-redpanda commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -1277,6 +1277,114 @@ def test_replication_with_truncated_topic(self):
+            )
+
+            # Produce a single message to ensure cluster linking picks up the trim
+            for part in range(0, 5):
```
theres still hardcoding of 5 in here

joe-redpanda commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -1277,6 +1277,114 @@ def test_replication_with_truncated_topic(self):
+                    )
+                    if (part.start_offset or 0) != offset:
+                        return False
+                return num_parts == 5 and True
```
Additional odd bool and True

joe-redpanda commented:
Test hardcoding / boolean case cleanup but otherwise LGTM

michael-redpanda committed:
cl/rep: Periodically synchronize all start offsets

Within link_replication_mgr, periodically sync all start offsets of all
partitions under management.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Fixed hardcoding of tests
* Moved periodic start offset syncing from partition_replicator to link_replication_mgr

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -245,4 +251,82 @@ ss::future<> partition_replicator::fetch_and_replicate() {
+ss::future<> partition_replicator::prefix_truncate(kafka::offset o) {
+    static constexpr auto prefix_truncate_timeout = 5s;
+    auto ec = co_await _sink->prefix_truncate(
+      o, ss::lowres_clock::now() + prefix_truncate_timeout);
```
handle exceptions? (I think unsetting of _in_progress_offset should happen in all cases)

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -245,4 +251,82 @@ ss::future<> partition_replicator::fetch_and_replicate() {
+      _in_progress_truncate_offset.has_value()
+      && o >= _in_progress_truncate_offset) {
+        _in_progress_truncate_offset = std::nullopt;
+    }
```
I think one simplification is to get rid of the queue, only have one in progress truncation, I feel that invariant simplifies a bunch of code here.  (spawn a fiber)

bharathv reviewed (commented):
handling exception is the main comment, some simplification suggestion, lgtm otherwise.

michael-redpanda commented:
Force push:

* Removed queue

michael-redpanda commented:
Force push:

* Removed unused dependency

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda commented:
Force push:

* Fixed error code check logic

michael-redpanda commented:
Force push:

* Fixed lint

michael-redpanda commented:
Forc epush:

* Use partition_proxy

bharathv commented on src/v/cluster_link/replication/partition_replicator.cc:
```diff
@@ -245,4 +246,80 @@ ss::future<> partition_replicator::fetch_and_replicate() {
+void partition_replicator::maybe_synchronize_start_offset() {
+    auto shadow_partition_hwm = _sink->high_watermark();
+    auto shadow_partition_start_offset = _sink->start_offset();
+    auto source_offsets = _source->get_offsets();
+
```
nit: move to lines below.

source_offsets after L261 and the rest two even lower.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74850-->
<!--5dcd2c17-3a03-44b5-905e-4f4a6744aebf-->
<!--e0ec141d-8aef-4a96-9bd0-b909c6e9ebc1-->
<!--33404607-ea3e-4b9a-8df7-bfcc6afa3ea8-->
<!--2ddcf78c-1514-463a-9a67-a7d84db898a8-->
<!--title-->
## Retry command for [Build#74850](https://buildkite.com/redpanda/redpanda/builds/74850)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

michael-redpanda commented:
rnot tests

michael-redpanda commented:
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74879-->
<!--0a3bb2a0-9914-432b-accd-72367a912c68-->
<!--f7829e86-cfb0-4ad8-91c8-41712dee6563-->
<!--title-->
## Retry command for [Build#74879](https://buildkite.com/redpanda/redpanda/builds/74879)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":false}
tests/rptest/tests/shadow_linking_rnot_test.py::ShadowLinkingRandomOpsTest.test_node_operations@{"failures":true}
```
<!--/body-->

michael-redpanda commented:
/ci-repeat 1
